#!/usr/bin/env python

# Script for copying the assemblies generated by Miguel's
# assembly script into the virome_contigs directory located
# in the Assemblies directory. It operates its commands from
# the Assemblies directory. First it checks if the contigs
# file is already present in the virome_contigs directory,
# if not then it copies the contigs.fasta file only if this
# exists.

import os
import subprocess

os.chdir("/hps/nobackup2/production/metagenomics/garp/Assemblies")


process = subprocess.Popen("ls -1 [ES]RP*/*/[ES]RR* | grep -x '[A-Z,0-9]\+'", stdout=subprocess.PIPE, shell=True)

run_files = process.stdout.read().decode("utf-8").split("\n")

del(run_files[-1])

for i in range(len(run_files)):
	run_files[i] = run_files[i] + ".fna"

for run_name in run_files:
	if run_name not in os.listdir("virome_contigs/"):
		run_name = run_name.rstrip(".fna")
		process = subprocess.Popen("find $(pwd) -type f | grep '%s[a-z,A-Z,0-9,/]\+contigs\.fasta'" % run_name, stdout=subprocess.PIPE, shell=True)
		file_path = process.stdout.read().decode("utf-8")
		if len(file_path) > 0:
			file_path = file_path.rstrip()
			subprocess.call("cp %s virome_contigs/%s.fna" % (file_path, run_name), shell=True)
